Webpack bundling implementation

PowerfulPony opened the pull request:
Project is divided into a demonstration part and an engine. Both parts are assembled using webpack. Possible dependencies have been updated and installed via npm. d3 has been removed from Material to reduce the size of the main module. In package.json, the "main" property specifies an UMD.

Future plans use ESLint with airbnb-base rules to format the entire code base in a single style.

I will be glad to criticism and comments, my main goal is to improve the usability of floom as a module, for use in other projects with various rendering systems.

Commit:
webpack bundling

Commit:
build 1.1.0 with demo

Commit:
Update README.md

PowerfulPony committed:
Merge pull request #1 from PowerfulPony/webpack

Webpack bundling implementation

PowerfulPony committed:
add editor config

PowerfulPony committed:
installed and configured eslint

PowerfulPony committed:
formatted demo

PowerfulPony committed:
linting floom

PowerfulPony committed:
Merge pull request #2 from PowerfulPony/eslint

Code linting by eslint added

onsetsu commented:
#### DISCLAIMER

To admit, I have some suspicion of malicious intent from your side, grounded in the emails you sent me prior to this pull request.
Empty email bodies with auto-translated subjects like *'Floom sehr cooles Projekt!'* look a lot like (rather personalized) spam mails and rose my suspicion.
As those are not a good basis to build trust, I'm more cautious about this than usual.

---

**Goals and Reasoning.** Your RP proposes some heavy changes to both the structure of the project and the workflow on adapting the project. Both come at a certain cost that has to be evaluated against the benefits.
A common vision in a team of contributors is essential for good teamwork. To be able to properly evaluate the proposed changes, they have to be put in a certain context. Therefore, it would be great to hear your goals and the reason for your interest in the project. Simply wanting to *'improve the usability of floom as a module'* seems like change for the sake of change, thus, is not a good motivator and key value.
I also saw [your work](https://github.com/PowerfulPony/fluid-simulation) on Grant Kot’s old web demo, proposing remarkably similar changes. However, these changes seem primarily technically motivated; probably to be able to apply your habitualized workflow on the project. However, technicalities should never be the starting point towards such a project.
Thus, I’m interested in the **reasoning and interest** behind those changes.

**Trade-offs and Habits.** As for the changes themselves, like I said, they bear a certain cost with them.
At the current point, everybody can just clone the project, look at it using any modern browser, and start making changes and experiment with it.
No further tooling required.
As someone who uses primarily self-sustaining and live programming environments, the ability to quickly perform experiments is key to my ideal programming experience.
I do not think there is a single “right” way do this project. Just trade-offs to be made.
Personally, I would want to preserve a low-friction way to work with this project, if possible.
Requiring contributors to use *npm* and *webpack* heavily impedes this core value of ease of use (in flavor of other benefits). Another disadvantage, due to the aforementioned suspicion of malicious intent, is that it opens up a big security hole in form of postinstall hooks.
In fact, a recent change I made was to remove all building steps entirely by removing a rollup step. This is actually not needed due to native support for es modules.

**Fix Problems, not Symptoms.** While it is true that the project needs some serious [reengineering]( http://scg.unibe.ch/download/oorp/OORP.pdf) due to its age, the proposed changes seem to fix symptoms rather than core problems. As an example, take the inclusion of lodash. Before, lodash was included as a whole while only a part of its functionality was used. The proposed change was to just include just `lodash.foreach`. While this change solves the intermediate symptom of an increased to download, it disregards the actual problem to be solved: lodash was included to accommodate for shortcomings in the JavaScript standard built-ins and language features. This, however, is not an issue anymore. Thus, the standard `forEach` with an anonymous function would render the inclusion of lodash unnecessary. Similarly, I see many of these ‘clockwork solutions’ that work across the board and do not require critical thinking and investigating (those are fine if you have to produce a lot of work, but usually do not present the ideal solution).

Maybe we should take this discussion offline, if you are still interested, that is. It might be a good opportunity to learn from each other.

PowerfulPony commented:
@onsetsu Thank you for your attention

I will try to explain in detail what I did and why I did it.


**Build using Webpack**

1) Helps you use absolute paths to modules (`../../` -> `@/`). This makes it easier to find and AutoCorrect dependencies, especially for young developing projects.
2) Babel transpilation-Allows you to use modern syntax in development, while maintaining the ability to use the project not only in modern browsers and interpreters.
3) Terser minimization and filtering-Minimizes the code without loss, allowing you to use the microband in demo projects that are built manually, increasing the size of the project to a lesser extent.
4) Webpack Analyzer - helps you clearly see dependencies, modules, and their relative sizes. Helps you track the weight of the module.
5) UMD bundle - Creates a module that works equally well in NODE.js and the Browser (including outdated versions).

I only listed what works now. Webpack allows you to easily perform other operations that may be required in the user, making a minimum of changes to the project's code base (webworkers, offset screen render, webassembly).
In addition, it makes it easy to generate demo pages and documentation without having to manually inject them.


**General stylistic design of the code**

1) .editorconfig Is the common basic settings for various code editors. the format is open and supported by a huge number of IDE's.
https://github.com/editorconfig/
https://editorconfig.org/
2) .gitattributes-Helps git tools look for changes more obviously. Also, some IDES rely on the rules from .gitattributes for automatic formatting.
3) ESLint-allows you to automatically search for syntactic errors, monitors the use and Declaration of variables, and can be used to adhere to the same style in writing code, both stylistically and ideologically.
4) eslint airbnb-base is a very hard, specific style guide for JS, contains a huge number of rules aimed at clarity, ease of debugging, and minimizing character changes in CVS. Each rule is quite clearly reasoned with examples.
https://github.com/airbnb/javascript

The main goal for style guides is a unified and predictable style for making changes. Depersonalization of the private style in favor of the General, obvious one. If a free library plans to develop at the expense of the community, and not in the person of a single author, then the adoption of General rules will help its development and inspection by other members of the community.


**Restructuring**

I saw two obvious fragments, the presentation part and the particle engine itself (MPM Fluid-Floom). The first one consists of various files and relies on using the render via the D3 library 2.0 version (which is easy to upgrade to 3.5). The second part is independent and deals with MPM calculation and calculation of circular collisions.
Using the build, I found that Floom.Material is linked to D3 (with the presentation part), which reduced the MPM bundle from ~200Kb to ~30Kb.
I also updated the dependent libraries and removed third-party code from the repository, except for the code that is no longer distributed via NPM.


**NPM and reliability** - for me as a front-end developer, NPM has proven to be a reliable source of packages and a system for optimizing dependencies.

Those who do not trust or support using their libraries without loading dependencies via NPM (or Yarn), each release adds ready-made UMD (or AMD, CMD) builds, minimized and not minimized. This measure, in my opinion, is excessive and practical, and it is also accepted by the community.


**My interest**

I am developing HTML5 games, usually I use them for physics matter.js. For the "Phaser", "PIXI", or CanvasRenderingContext2D render. Now I am create to make a game similar to "PixelJunk Shooter". My goal is to make a browser implementation and weighing less than 1Mb, I could pay attention to Box2d and LiquidFun (https://google.github.io/liquidfun/) but these are very heavy libraries. I turned my attention to MPM Fluid (https://github.com/kotsoft/MPM-Fluid) and in General on MPM, practice has shown that this solution provides good fluid behavior and acceptable operation of ~1000 particles at a time.
There are enough examples on the Internet where the MPM approach is used and the render is implemented, but there is not a single developed solution in JS other than yours. But your solution now looks more like a technological demonstration. To use it in my project, I would have to delete the demo layer as not needed. But in this case, I would lose a simple update option.
And I am sure you will continue to develop the project and I would not like to lose your updates. In my opinion, what is cool, and what can be, is beyond the bounds of possibility!


ps/ My PR is open to changes, and I am always open to dialogue.
